### ISSUE TITLE
Collection elements should not just be hidden, but use a distinct state

### ISSUE THREAD

**Issue by wm75** [galaxyproject/galaxy#10896]
Collections are perfect to  use with complex workflows except for one major annoyance:

Not making a collection a WF output currently is a *really* bad idea cause a hidden collection puts you in debugging hell if something goes wrong with your WF. That's because to then find a collection again and access its datasets you'd have to unhide *all* hidden datasets.

Though I don't have a perfect solution for that problem, my preferred one would be to introduce an additional dataset state (on top of hidden and deleted) which would indicate dataset_is_part_of_collection. Now when you'd unhide hidden datasets that would *not* reveal any collection elements.

**Comment by dannon** [comment]
So this is if a single job in a collection fails?  What if the collection itself just didn't hide if any individual datasets were in non-okay states?

**Comment by wm75** [comment]
That would work in a subset of situations, yes.
However, you may also have a downstream step fail, but you suspect the real issue is at the collection step before.

I think, this issue isn't really tied to job failures in collections (it's just the situation I experience most often), but more about accessibility of hidden datasets, in general.
The very reason why you may want to use collections is because you have *many* datasets that should be processed in parallel.
Collections are great in these situations not least because they allow you to produce clean histories that are manageable even though each collection may represent hundreds of datasets. Just that right now you completely lose this benefit as soon as you have to inspect hidden data.

**Comment by wm75** [comment]
Oh, and I'm aware that Advanced history search can help here a bit cause you could filter the hidden datasets using `history_content_type=collection`. However this is:
- relatively slow for really big histories
- hard to remember (I have to look it up nearly every time)
- almost undiscoverable except for very experienced users

**Comment by mvdbeek** [comment]
> which would indicate dataset_is_part_of_collection. Now when you'd unhide hidden datasets that would _not_ reveal any collection elements.

How about making the `unhidden` link a two-step thing (all - dataset collections only)?
That should be straightforward to add.
<img width="190" alt="Screenshot 2020-12-11 at 10 19 37" src="https://user-images.githubusercontent.com/6804901/101885548-8eb28a80-3b9a-11eb-949e-5fe98f0a74eb.png">

**Comment by wm75** [comment]
Hmm, that's a nice idea! Conceptually, I still think that (ab)using the hidden state to get collection elements out of the way is not the cleanest solution, *but* it seems your suggestion really is *much* easier to implement.
How performant would this be on large histories though? I imagine you would still have to iterate over all hidden datasets to filter out the collections?

**Comment by mvdbeek** [comment]
> , I still think that (ab)using the hidden state to get collection elements out of the way is not the cleanest solution, _but_ it seems your suggestion really is _much_ easier to implement.

I agree it would be nice if we could skip the `History` part in HistoryDatasetAssociation, so the relationship could become `HistoryDatasetCollectionAssociation` -> `DatasetCollection` -> `DatasetCollectionElement` -> `DatasetAssociation` (instead of `HistoryDatasetAssociation`) -> `Dataset`, but there's a lot of details to work out, like how do you calculate the history quota if the DatasetAssociation is not tied to a history, and there's a ton of places in the code where we infer the user (and other things) via the history relationship. But there's also a lot of advantages to be had if we did this.



> How performant would this be on large histories though? I imagine you would still have to iterate over all hidden datasets to filter out the collections?

That would be in the SQL query, which is distinct for for HDCAs and HDAs, so I don't think this would be much slower than the current query that only includes visible HDCAs and HDAs.

**Comment by wm75** [comment]
Yeah, I can see that this is far from trivial and your simpler UI solution should also just fix my issue :)
